mio/src/lib.rs

```diff
@@ -115,77 +149,128 @@ impl TcpStream {
                     let len = into.len();
                     into.set_len(len + i);
                 }
-                return futures::finished(into).boxed()
+                Some(Ok(into))
             }
             Err(e) => {
-                if e.kind() != io::ErrorKind::WouldBlock {
-                    return futures::failed(Error::new(e, into)).boxed()
+                if e.kind() == io::ErrorKind::WouldBlock {
+                    self.into = Some(into);
+                    None
+                } else {
+                    Some(Err(PollError::Other(Error::new(e, into))))
                 }
             }
         }
-        let (p, c) = promise();
-        let r = self.tx.send(Message::Wait(c,
-                                           mio::EventSet::readable(),
-                                           self.tcp.clone()));
-        match r {
-            Ok(()) => {
-                let me2 = TcpStream {
-                    tcp: self.tcp.clone(),
-                    tx: self.tx.clone(),
-                };
-                p.then(move |res| {
-                    match res {
-                        Ok(()) => me2.read(into),
-                        Err(e) => {
-                            futures::failed(Error::new(e, into)).boxed()
-                        }
-                    }
-                }).boxed()
-            }
-            Err(mio::channel::SendError::Io(e)) => {
-                return futures::failed(Error::new(e, into)).boxed()
-            }
-            Err(mio::channel::SendError::Disconnected(..)) => panic!("closed channel"),
-        }
     }
+}
+
+pub struct WriteCompletion {
```

Comment:
You probably feel the same way, but as many of these as possible should probably be private